### PR TITLE
Record location in rust log checks

### PR DESCRIPTION
**Pull request by rmcgibbo**
Fix https://github.com/jtojnar/nixpkgs-hammering/issues/94

### CODE REVIEW

File: rust-checks/src/bin/no-python-tests.rs [hunk 1]
```diff
@@ -11,17 +11,21 @@ fn main() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
-fn analyze_single_file(log: BufReader<ChildStdout>) -> Result<Report, Box<dyn Error>> {
-    let re = Regex::new(r"Ran 0 tests in \d+.\d+s").unwrap();
+fn analyze_single_file(log: BufReader<ChildStdout>, attr: &Attr) -> Result<Report, Box<dyn Error>> {
+    let re = Regex::new(r"(Ran 0 tests|no tests ran\x1b\[0m\x1b\[33m) in \d+.\d+s").unwrap();
```
Comment (jtojnar) [inline rust-checks/src/bin/no-python-tests.rs:15]: Could you add a comment explaining what each alternative is for?

File: rust-checks/src/bin/no-python-tests.rs [hunk 2]
```diff
@@ -11,17 +11,37 @@ fn main() -> Result<(), Box<dyn Error>> {
...
         .filter_map(|s| re.find(&s).and_then(|m| Some(m.as_str().to_string())))
         .map(|m| NixpkgsHammerMessage {
-            msg: format!("Test runner could not discover any test cases: ‘{}’", m),
+            msg: format!(
+                "Test runner could not discover any test cases: ‘{}’",
+                ansi_escape.replace(&m, "")
+            ),
             name: "no-python-tests",
-            locations: vec![],
+            locations: attr.location.iter().map(|x| x.clone()).collect(),
```
Comment (rmcgibbo) [inline rust-checks/src/bin/no-python-tests.rs:44]: One limitation of this is that it doesn't really flag the ideal _line_ within the nix expression for the attr. e.g.

```
nix run -f ~/projects/nixpkgs-hammering -c nixpkgs-hammer -f . python38Packages.humanize -e attribute-orderingWhen evaluating attribute ‘python38Packages.humanize’:
warning: no-python-tests
Test runner could not discover any test cases: ‘no tests ran in 0.00s’
Near /home/mcgibbon/projects/nixpkgs/pkgs/development/python-modules/humanize/default.nix:30:0:
   |
30 |     description = "Python humanize utilities";
   | ^
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
```

File: tools/nixpkgs-hammer [hunk 3]
```diff
@@ -206,7 +206,7 @@ def render_code_location_ansi(loc: SourceLocation) -> str:
         pointer = " " * (column - 1) + "^"
 
         location_lines = [
-            f"Near {loc.file}:{loc.line}:{loc.column}:",
+            f"Near {loc.file}:{loc.line}:{loc.column if loc.column is not None else 0}:",
```
Comment (jtojnar) [inline tools/nixpkgs-hammer:209]: Maybe if we do not know the column, we should just omit the `:0`.

File: tools/nixpkgs-hammer [hunk 4]
```diff
@@ -203,7 +203,7 @@ def render_code_location_ansi(loc: SourceLocation) -> str:
         pointer = " " * (column - 1) + "^"
 
         location_lines = [
-            f"Near {loc.file}:{loc.line}:{loc.column}:",
+            f"Near {loc.file}:{loc.line}:",
```
Comment (jtojnar) [inline tools/nixpkgs-hammer:206]: Is this intentional?

Comment (rmcgibbo) [reply tools/nixpkgs-hammer:206]: Oh whoops, that's not what I want. One sec.

Comment (rmcgibbo) [reply tools/nixpkgs-hammer]: Okay,  fixed up better.

File: tools/nixpkgs-hammer [hunk 5]
```diff
@@ -201,9 +201,14 @@ def render_code_location_ansi(loc: SourceLocation) -> str:
         line_spaces = " " * len(str(loc.line))
         column = loc.column if loc.column is not None else 1
         pointer = " " * (column - 1) + "^"
+        rendered_loc = ":".join(
+            (loc.file, str(loc.line)) + (str(loc.column),)
+            if loc.column is not None
+            else ()
```
Comment (jtojnar) [inline tools/nixpkgs-hammer:207]: This reads wrong to me.

And apparently, Python uses this wrong association:

```
>>> 3 + 2 if False else 1
1
```

Comment (rmcgibbo) [reply tools/nixpkgs-hammer]: I rewrote it to be procedural.

**Comment by jtojnar** [comment]
Thanks.

**Pull request merged**